File: BinderServicemanager/java/TestServer.java
/**
 * 1. add Service
 * 2. while(true) { read data, prase data, call function, reply}
 */
import android.util.Slog;
import android.os.ServiceManager;

public class TestServer {
	private static final String TAG = "TestServer";

	public static void main(String[] args) {
		/* add Service */
		Slog.i(TAG,"add Hello Service");
		ServiceManager.addService("hello", new HelloService() );

		while(true)
		{
			try
			{
				Thread.sleep(100);
			}
			catch( Exception e )
			{

			}
		}
	}
}

File: BinderServicemanager/cpp/test_server.cpp
/* 参考： frameworks/av/media/mediaserver/main_mediaserver.cpp  */

#include "IHelloService.h"
#include "IGoodbyeService.h"

#include <fcntl.h>
#include <sys/prctl.h>
#include <sys/wait.h>
#include <binder/IPCThreadState.h>
#include <binder/ProcessState.h>
#include <binder/IServiceManager.h>
#include <cutils/properties.h>
#include <utils/Log.h>


using namespace android;


int main(int argc, char const *argv[])
{
	/* addService */

	/* loop: read data, prashe data , call function */

	/* 打开驱动 */
	sp<ProcessState> proc(ProcessState::self());
	ALOGI("open device ");
	/* 获得BpServiceManager */
    sp<IServiceManager> sm = defaultServiceManager();

    sm->addService(String16("hello"), new BnHelloService());
    sm->addService(String16("goodbye"), new BnGoodbyeService());

    /* 循环体 */
    ProcessState::self()->startThreadPool();
    IPCThreadState::self()->joinThreadPool();

    return 0;

}
